### PR TITLE
Add TCP protocol support to Socket

### PR DESCRIPTION
**Pull request by pekiZG**
Added TCP protocol parameter switch.
Name refactoring.

### CODE REVIEW

**Comment by DarrellMozingo** [comment]
Thanks for the PR Petar! A few concerns:

1. The project .NET version being upgraded is a breaking change we weren't ready to make yet. What forced you to do it?
2. No smoke/acceptance tests for the TCP functionality itself
3. I don't see the TCP connection being opened anywhere? According to the docs for [`SendTo()`](https://msdn.microsoft.com/en-us/library/w5zdhc3x(v=vs.110).aspx#Anchor_2), it should be throwing an exception:

>  If you are using a connection-oriented protocol, you must first establish a remote host connection by calling the Connect method or accept an incoming connection request using the Accept method. If you do not establish or accept a remote host connection, SendTo will throw a SocketException.

The management of a TCP connection is why I was hesitant to add this functionality in the first place. I can't see a full handshake being done before each send request since it's all blocking and would kill the calling app. It's all fully possible of course, just more than I wanted to add for something I still don't see a huge amount of benefit from :)

**Comment by pekiZG** [comment]
Hi Darrell,
This is my first PR :)

I would like to address all your concerns.

Break was not intentional. Visual Studio 2015 on Windows 10 wouldn't open it.
Later I found how to enable .net 2.0+ and will migrate it without breaking.

2. Still looking why tests won't run. It's probably connected to bullet number 1.

3. I tried it and it didn't throw an exception. SendTo() should be used on line 107 in StatsdUDP.cs
I'm not happy how I did it.

**Comment by pekiZG** [comment]
@DarrellMozingo 

I have successfully run all the tests after poking around and finding out that NUnit 2.6 is being used.
Will have TCP connection test soon.


KR
Petar

**Comment by pekiZG** [comment]
After finding out that there is App.confing in Tests I have created Hyper-V and configured InfluxDB with Chronograph and Telegraph to have valid Smoke tests. (See picture below)

image

TCP Smoke test is passing.

I have changed to the implementation towards the interface over concrete type.
You can see it in the last commit (src/StatsdClient/Metrics.cs)

Also, I'm not done yet.

Thinking about doing it with async methods. @DarrellMozingo that should address your concern on blocking the calling app. Right?

**Comment by DarrellMozingo** [comment]
Thanks for fixing the .NET version and splitting out the TCP/UDP classes - nice touch.

I see you're doing the connect/disconnect now. I think that should be in a try/finally so ensure the connection is closed if something goes wrong in the send (I'll add an in-line comment). Also a base class (or a separate class) to handle the IP address resolution as it's a bit convoluted?

I've personally always been against doing the TCP for socket management and the speed hit of doing a sync call. If it's an absolute requirement, I feel it's better to split that responsibility out to another app (ie have a statsd relay running on each server that you'd send UDP stats to, and it would relay them in TCP). Do you still think the feature is needed if someone does this?

Thanks again for taking the time to contribute, I do appreciate it! :)

File: src/StatsdClient/StatsdTCP.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        public void Send(string command)
+        {
+            Send(Encoding.ASCII.GetBytes(command));
+        }
+
+        private void Send(byte[] encodedCommand)
+        {
+            _clientSocket.Connect(IPEndpoint.Address, _port);
+            _clientSocket.SendTo(encodedCommand, encodedCommand.Length, SocketFlags.None, IPEndpoint);
+            _clientSocket.Disconnect(false);
```
Comment (DarrellMozingo) [inline src/StatsdClient/StatsdTCP.cs]: Put this in a `try {} finally {}` so we're not leaving hanging TCP connections?

File: src/StatsdClient/StatsdTCP.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            _name = name;
+            _port = port;
+
+            _clientSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+
+            var ipAddress = GetIpv4Address(name);
+            IPEndpoint = new IPEndPoint(ipAddress, _port);
+        }
+
+        private IPAddress GetIpv4Address(string name)
```
Comment (DarrellMozingo) [inline src/StatsdClient/StatsdTCP.cs]: Maybe pull this out to a separate class so both TCP & UDP implementations can share it?

**Comment by pekiZG** [comment]
@DarrellMozingo no problem I really want to contribute.
Also, I will do a little How to setup development environment. Recently had to do a clean install and it took me time to setup.

My position is that you should have a choice between UDP and TCP.
If someone has control over the network and can place another app. that sounds like a better option.

Today I will work on this part from documentation.
https://msdn.microsoft.com/en-us/library/system.net.sockets.socket(v=vs.80).aspx
> To process communications using separate threads during execution, use the following methods, which are designed for asynchronous operation mode

Using BeginSend and EndSend would be much better.
I will try to make some kind of load test. Have no idea how (actually I have).
Have to think about what I want to measure.

On the topic of IP address resolution, that should definitely be pulled out (separation of concern).


I really appreciate you taking time to look at my messy code :)
Pair programming :+1:

File: src/StatsdClient/StatsdTCP.cs [hunk 3]
```diff
@@ -54,9 +25,16 @@ public void Send(string command)
...
         private void Send(byte[] encodedCommand)
         {
-            _clientSocket.Connect(IPEndpoint.Address, _port);
-            _clientSocket.SendTo(encodedCommand, encodedCommand.Length, SocketFlags.None, IPEndpoint);
-            _clientSocket.Disconnect(false);
+            try
+            {
+                _clientSocket.Connect(IPEndpoint);
+                _clientSocket.SendTo(encodedCommand, encodedCommand.Length, SocketFlags.None, IPEndpoint);
+                _clientSocket.Disconnect(false);
```
Comment (DarrellMozingo) [inline src/StatsdClient/StatsdTCP.cs]: The `Disconnect()` should be in the `finally` block, no?

File: src/StatsdClient/AdressResolution.cs [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+            if (!isValidIpAddress)
+            {
+                ipAddress = GetIpFromHostname(name);
+            }
+
+            return ipAddress;
+        }
+
+        public IPAddress GetIpFromHostname(string name)
```
Comment (DarrellMozingo) [inline src/StatsdClient/AdressResolution.cs]: Can be made private I think?

**Comment by DarrellMozingo** [comment]
I'm still hesitant to add it, but I suppose the TCP code will be insulated in a separate class and not defaulted or anything, so we can allow it with perhaps appropriate warnings (ie to use UDP or a TCP relay app first).

Thanks for the fixes so far! I'll be away for a few days, so if you get anything else done I'll take a look next week.

File: src/StatsdClient/StatsdTCP.cs (hunk 3 above)
Comment (pekiZG) [inline src/StatsdClient/StatsdTCP.cs]: Yes,

still forking on this part to make it Async. :)

**Comment by pekiZG** [comment]
I haven't had time over the weekend.
I will have it finished tonight \o/

File: src/StatsdClient/AdressResolution.cs (hunk 4 above)
Comment (pekiZG) [inline src/StatsdClient/AdressResolution.cs]: Private, i don't think so... maybe protected?

File: src/StatsdClient/StatsdTCP.cs [hunk 5]
```diff
@@ -35,16 +34,15 @@ private void Send(byte[] encodedCommand)
...
-                _clientSocket.Close();
             }
             catch (Exception ex)
             {
-                Trace.TraceError(ex.Message);
+                Console.WriteLine(ex.Message);
             }
             finally
             {
-                _clientSocket.Disconnect(false);
```
Comment (pekiZG) [inline src/StatsdClient/StatsdTCP.cs]: Removed Disconnect() in favor of Shutdown() and Close()

**Comment by pekiZG** [comment]
@DarrellMozingo I think I'm done. \o/
Need second opinion.

**Comment by DarrellMozingo** [comment]
Nice work!

I know `IDisposable` was already on the UDP version, but seeing it do nothing in the TCP one and not really being called anywhere in the UDP version makes me wonder if it's even needed?

Also can you rename the TCP smoke test to match the class/other UDP one?

**Comment by pekiZG** [comment]
Hi @DarrellMozingo

I have implemented IDisposable interface, and renamed filenames to reflect classnames.
I didn't catch what you mean to rename smoketests?

Is there some convention I'm not being aware of?

also, I'm logging to console?
Do you think that's a smart thing to do? Or just to "swallow" the error?

**Comment by DarrellMozingo** [comment]
Yea, saw you implmented `IDisposable`. I was wondering if it could be removed all together? I don't see `Dispose()` being called on the UDP version. Maybe it's best to just close the UDP socket inside the `Send()` method as well...

Ignore my smoke test naming comment. I misread something and thought the UDP test was named differently. It's fine as is.

Swallowing seems wrong. I think something like [TraceSource](https://msdn.microsoft.com/en-us/library/system.diagnostics.tracesource(v=vs.110).aspx) (built into .NET) or [LibLog](https://github.com/damianh/LibLog) would be best, but we're not doing either now and there hasn't been a need yet. The console log seems good enough for now anyway :)

**Comment by DarrellMozingo** [comment]
The build for this branch is [broken](https://ci.appveyor.com/project/DarrellMozingo/statsd-csharp-client/build/1.2.38). Not sure why the PR didn't update (I just enabled it, maybe I did it wrong!).

Looks like the project file still has `StatsdTCP.cs` instead of `StatsdTCPClient.cs` (same for UDP as well)?

**Comment by pekiZG** [comment]
well, thinking is that if you close the outside connection (finally block), you shouldn't worry about dispose.

Or if you are doing something like this:
https://msdn.microsoft.com/en-us/library/yh598w02(v=vs.140).aspx
```
using(mySocket)
{
// do work
}
```
Dispose will be properly called.

----
Yes, I changed the names; Didn't know what you were referencing on SmokeTest

**Comment by pekiZG** [comment]
Yes, you are right... i will fix it now :)

**Comment by DarrellMozingo** [comment]
Regarding disposable, you're absolutely correct. No body is doing a `using()` with it though, and you're disposing the socket right away in the TCP class. Perhaps we should do the same with the UDP one then remove the `IDisposable` interface entirely? It would not be doing anything then.

**Comment by pekiZG** [comment]
Here is what happened.
Didn't push the changed StatsdClient.csproj

**Comment by pekiZG** [comment]
Yes, TCP smoke test will fail.
UDP will not because it's not waiting on connection.

I should change the smoke test to use TcpListener

**Comment by pekiZG** [comment]
@DarrellMozingo 

https://ci.appveyor.com/project/DarrellMozingo/statsd-csharp-client/build/tests

What should we do with TCP SmokeTest?

**Comment by DarrellMozingo** [comment]
I like the idea of the smoke test spinning up a simple TcpListener to receive the connection. As a bonus it can read the data and make sure it's correct :)

**Comment by pekiZG** [comment]
I was under Impression that Smoke test should be done on 'real' thing :)

That's why I did all [this](http://petar-slat.from.hr/2016/04/25/how-to-setup-influxdb-chronograf-and-telegraf-for-development/)

**Comment by DarrellMozingo** [comment]
You're correct, it should be hitting the real system. From the point of view of the client though I can see just making sure it's sending a properly formed TCP packet, rather than a full end-to-end test that makes sure it shows up in the back-end correctly. So just spinning up a TCP socket to send to should work well enough for our purposes.

**Comment by pekiZG** [comment]
@DarrellMozingo 1.2.40 built successfully

https://ci.appveyor.com/project/DarrellMozingo/statsd-csharp-client/build/artifacts

We should probably update the readme.md :+1:

File: src/Tests/TCPSmokeTests.cs [hunk 6]
```diff
@@ -9,21 +9,41 @@ namespace Tests
...
         [Test]
         public void Sends_a_counter()
         {
+			TcpListener server = null; 
+			// Set the TcpListener
+			Int32 port = _serverPort;
+			System.Net.IPAddress localAddr = System.Net.IPAddress.Parse(_serverName);
+
+			// TcpListener server = new TcpListener(port);
+			server = new TcpListener(localAddr, port);
```
Comment (DarrellMozingo) [inline src/Tests/TCPSmokeTests.cs]: Can you wrap this in a `using{}` to make sure it's always stopped?

**Comment by DarrellMozingo** [comment]
Looks good, thanks! Want to take a stab at updating the README?

**Comment by pekiZG** [comment]
Sure,

I would love to :)

File: src/Tests/TCPSmokeTests.cs (hunk 6 above)
Comment (pekiZG) [inline src/Tests/TCPSmokeTests.cs]: TcpListener class doesn't inherit/implement IDisposable Interface.
For that reason It's can't be wrapped in using()

If you try you get _Type used in a using statement must be implicitly convertible to 'System.IDisposable'_

Comment (DarrellMozingo) [inline src/Tests/TCPSmokeTests.cs]: Quite right :) I was thinking of `TcpClient`, sorry. Maybe just put the `Stop()` in a finally block, or probably a test teardown fixture would be better. Just want to make sure if the test fails that listening socket gets closed.

Comment (pekiZG) [inline src/Tests/TCPSmokeTests.cs]: I went with setup/teardown fixture :)

**Comment by pekiZG** [comment]
@DarrellMozingo readme is updated :)

The changelog is left up to you :+1:

**Comment by DarrellMozingo** [comment]
Great, thanks a lot for the contribution Petar! I'll get it pushed out today or tomorrow.

**Comment by pekiZG** [comment]
No hurry :)

I'm here if additional tweaks are needed.

**Pull request merged**